rio_tiler/io/base.py

```diff
@@ -108,6 +109,103 @@ def tile_exists(self, tile_z: int, tile_x: int, tile_y: int) -> bool:
         )
 
 
+@attr.s
+class AsyncBaseReader(metaclass=abc.ABCMeta):
+    """Rio-tiler.io AsyncBaseReader."""
+
+    tms: TileMatrixSet = attr.ib(default=WEB_MERCATOR_TMS)
+    bounds: Tuple[float, float, float, float] = attr.ib(init=False)
+    minzoom: int = attr.ib(init=False)
+    maxzoom: int = attr.ib(init=False)
+
+    async def __aenter__(self):
+        """Support using with Context Managers."""
+        return self
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        """Support using with Context Managers."""
+        pass
+
+    @property
+    def center(self) -> Tuple[float, float, int]:
+        """Dataset center + minzoom."""
+        return (
+            (self.bounds[0] + self.bounds[2]) / 2,
+            (self.bounds[1] + self.bounds[3]) / 2,
+            self.minzoom,
+        )
+
+    @abc.abstractmethod
+    async def info(self) -> Coroutine[Any, Any, Dict]:
+        """Return Dataset's info."""
+        ...
+
+    @property
+    def spatial_info(self) -> Dict:
+        """Return Dataset's spatial info."""
+        return {
+            "bounds": self.bounds,
+            "center": self.center,
+            "minzoom": self.minzoom,
+            "maxzoom": self.maxzoom,
+        }
+
+    @abc.abstractmethod
+    async def stats(
+        self, pmin: float = 2.0, pmax: float = 98.0, **kwargs: Any
+    ) -> Coroutine[Any, Any, Dict]:
+        """Return Dataset's statistics."""
+        ...
+
+    async def metadata(
+        self, pmin: float = 2.0, pmax: float = 98.0, **kwargs: Any,
+    ) -> Coroutine[Any, Any, Dict]:
+        """Return Dataset's statistics and info."""
+        info, stats = await asyncio.gather(
+            *[self.info(), self.stats(pmin, pmax, **kwargs)]
+        )
+        info["statistics"] = stats
+        return info
+
+    @abc.abstractmethod
+    async def tile(
+        self, tile_x: int, tile_y: int, tile_z: int, **kwargs: Any
+    ) -> Coroutine[Any, Any, Tuple[numpy.ndarray, numpy.ndarray]]:
+        """Read a Map tile from the Dataset."""
+        ...
+
+    @abc.abstractmethod
+    async def part(
+        self, bbox: Tuple[float, float, float, float], **kwargs: Any
+    ) -> Coroutine[Any, Any, Tuple[numpy.ndarray, numpy.ndarray]]:
+        """Read a Part of a Dataset."""
+        ...
+
+    @abc.abstractmethod
+    async def preview(
+        self, **kwargs: Any
+    ) -> Coroutine[Any, Any, Tuple[numpy.ndarray, numpy.ndarray]]:
+        """Return a preview of a Dataset."""
+        ...
+
+    @abc.abstractmethod
+    async def point(
+        self, lon: float, lat: float, **kwargs: Any
+    ) -> Coroutine[Any, Any, List]:
+        """Read a value from a Dataset."""
+        ...
+
+    def tile_exists(self, tile_z: int, tile_x: int, tile_y: int) -> bool:
```

Comment:
This could go into a mixin if we wanted to dedup with `BaseReader`

Reply:
not sure to understand this @geospatial-jeff

Reply:
You can have a `TileExistsMixin` that defines `tile_exists` (and maybe `center` and `spatial_info`?) since those are the same across the `BaseReader` and the `AsyncBaseReader`. Then you add the mixin to both classes.

If you want to make the mixin 👍 , but not a strong preference.

Reply:
Just that this method is the same across the sync and async base classes.  We could only define this method in one place by writing a mixin which is shared across both base classes.  But repeating code isn't that big a deal so I'm fine leaving it as is as well!

Reply:
the Mixin is a great idea! I'll make a PR and then we can merge this